Caio Madeira e Guilherme Cavazotto

In [7]:
import pandas as pd 
import matplotlib.pyplot as plt
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split

In [8]:
df = pd.read_csv("cardio_train.csv", sep=";")
df.head()

,id,age,gender,height,weight,ap_hi,ap_lo,cholesterol,gluc,smoke,alco,active,cardio
0,0,18393,2,168,62.0,110,80,1,1,0,0,1,0
1,1,20228,1,156,85.0,140,90,3,1,0,0,1,1
2,2,18857,1,165,64.0,130,70,3,1,0,0,0,1
3,3,17623,2,169,82.0,150,100,1,1,0,0,1,1
4,4,17474,1,156,56.0,100,60,1,1,0,0,0,0


Construa um subset desse dataset, usando 1000 amostras da classe 0 (doença ausente) e 1000 amostras da classe 1 (doença presente).  Use 80% para o dataset de treino e 20% para o dataset de teste.

In [9]:
# Confesso que eu ainda não entendi o motivo de separar 1000 amostras de cada. 
# Custo computacional? PS: Eu preciso juntar depois, saquei.
df_cls0 = df[df['cardio'] == 0].sample(n=1000, random_state=42)
df_cls1 = df[df['cardio'] == 1].sample(n=1000, random_state=42)

Nota: entendo que X (maiusculo) são minhas features. O que tenho/caracteristicas/dados de entrada. y é o rotulo e variavel alvo. O que eu quero prever.

X_train e y_train: dados de treino. É os dados pra aprendizado + gabarito. Ele vai olhar pros dados de treino X_train e pras respostas y_train.

X_test são os dados de teste. Depois que o modelo treino, ele recebe o X_test que sao apenas as perguntas sem respostas e com isso ele preve em cima. O y_test seria o gabarito que fica escondido. Quando o modelo devolver as previsoes feitas em X_test, eu comparo essas previsoes com o y_test verdadeiro. Essa etapa é o calculo da acuracia ou erro.

In [10]:
df_sub = pd.concat([df_cls0, df_cls1])

X = df_sub.drop(columns=['cardio', 'id']) # tudo menos o que quero prever e os ids(ruido)
y = df_sub['cardio'] # o que quero prever

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.20,
    stratify=y,
    random_state=42,
)

In [11]:
print(X_train.values)
print(y_train)

[[1.6959e+04 1.0000e+00 1.6200e+02 ... 0.0000e+00 0.0000e+00 1.0000e+00]
 [1.9856e+04 1.0000e+00 1.6400e+02 ... 0.0000e+00 0.0000e+00 0.0000e+00]
 [2.1314e+04 2.0000e+00 1.7500e+02 ... 0.0000e+00 0.0000e+00 1.0000e+00]
 ...
 [2.0635e+04 1.0000e+00 1.6800e+02 ... 0.0000e+00 0.0000e+00 0.0000e+00]
 [2.1971e+04 2.0000e+00 1.8500e+02 ... 0.0000e+00 0.0000e+00 0.0000e+00]
 [2.2110e+04 1.0000e+00 1.6200e+02 ... 0.0000e+00 0.0000e+00 1.0000e+00]]
28841    1
1287     1
6096     0
36800    0
25758    1
        ..
53629    1
27059    0
3851     1
53100    1
16352    1
Name: cardio, Length: 1600, dtype: int64


In [12]:
for k in range(2, 10):
    knn = KNeighborsClassifier(n_neighbors=k)
    #knn.fit(X_train, y_train)
    knn.fit(X_train.values, y_train.values) # se eu uso o .values eu apro de receber warning
    # aparentemente eh mais  facil treinar o modelo usando numeros puros sem strs
    print(f"K={k}; Acc={knn.score(X_test.values, y_test)}")

    correctness = 0
    for i in range(0, len(X_test)): # enbtendo por X_test os novos "data points"
       # print("x_test.values=", [X_test.values[i]])
        predict = knn.predict([X_test.values[i]])
        #print(X_test.values[i], "prediction: ", predict)
        if predict == y_test.iloc[i]: # tive um problema de indexação entao troquei p .iloc[i]
            correctness += 1
    acc = correctness / len(X_test)
    print("acertos=", correctness)
    print("erros=", len(X_test) - correctness)
    print("accuracy=", acc)

K=2; Acc=0.585
acertos= 234
erros= 166
accuracy= 0.585
K=3; Acc=0.5825
acertos= 233
erros= 167
accuracy= 0.5825
K=4; Acc=0.575
acertos= 230
erros= 170
accuracy= 0.575
K=5; Acc=0.5875
acertos= 235
erros= 165
accuracy= 0.5875
K=6; Acc=0.5675
acertos= 227
erros= 173
accuracy= 0.5675
K=7; Acc=0.5925
acertos= 237
erros= 163
accuracy= 0.5925
K=8; Acc=0.585
acertos= 234
erros= 166
accuracy= 0.585
K=9; Acc=0.595
acertos= 238
erros= 162
accuracy= 0.595


Aparentemente o melhor resultado foi com k=9 com 59% de acc com maior numero absoluto de acertos.